File: csharp/Behavioral/22-Visitor/Visitor/ComplianceAuditVisitor.cs
namespace Visitor;

public class ComplianceAuditVisitor : IJobVisitor
{
    private readonly List<string> _issues = new();
    private int _checked = 0;

    public void Visit(PlumbingJob job)
    {
        _checked++;

        if (job.PressureBar > 4m)
            _issues.Add($"#{job.JobId} — pressure {job.PressureBar} bar exceeds " +
                        $"SANS 10400-W maximum of 4 bar");
    }

    public void Visit(ElectricalJob job)
    {
        _checked++;

        if (job.CircuitLoadAmps > 60m && !job.RequiresCertificate)
            _issues.Add($"#{job.JobId} — {job.CircuitLoadAmps}A load requires a " +
                        $"Certificate of Compliance, none flagged");
    }

    public void Visit(CarpentryJob job)
    {
        _checked++;

        if (job.FinishType == "untreated")
            _issues.Add($"#{job.JobId} — untreated timber not permitted in wet areas");
    }

    public void PrintReport()
    {
        Console.WriteLine("\n=== Compliance Audit ===");
        Console.WriteLine($"  Jobs checked : {_checked}");
        Console.WriteLine($"  Issues found : {_issues.Count}");

        foreach (var issue in _issues)
            Console.WriteLine($"  ⚠️  {issue}");

        if (_issues.Count == 0)
            Console.WriteLine("  ✅ All jobs compliant.");
    }
}

File: typescript/Behavioral/22-Visitor/Visitor/ComplianceAuditVisitor.ts
import type { IJobVisitor } from "./IJobVisitor";
import type { PlumbingJob } from "../Element/PlumbingJob";
import type { ElectricalJob } from "../Element/ElectricalJob";
import type { CarpentryJob } from "../Element/CarpentryJob";

export class ComplianceAuditVisitor implements IJobVisitor {
  private issues: string[] = [];
  private checked: number = 0;

  visitPlumbing(job: PlumbingJob): void {
    this.checked++;

    if (job.pressureBar > 4)
      this.issues.push(
        `#${job.jobId} — pressure ${job.pressureBar} bar exceeds ` +
          `SANS 10400-W maximum of 4 bar`,
      );
  }

  visitElectrical(job: ElectricalJob): void {
    this.checked++;

    if (job.circuitLoadAmps > 60 && !job.requiresCertificate)
      this.issues.push(
        `#${job.jobId} — ${job.circuitLoadAmps}A load requires a ` +
          `Certificate of Compliance, none flagged`,
      );
  }

  visitCarpentry(job: CarpentryJob): void {
    this.checked++;

    if (job.finishType === "untreated")
      this.issues.push(
        `#${job.jobId} — untreated timber not permitted in wet areas`,
      );
  }

  printReport(): void {
    console.log("\n=== Compliance Audit ===");
    console.log(`  Jobs checked : ${this.checked}`);
    console.log(`  Issues found : ${this.issues.length}`);

    for (const issue of this.issues) console.log(`  ⚠️  ${issue}`);

    if (this.issues.length === 0) console.log("  ✅ All jobs compliant.");
  }
}
